File: streamer-v2/src/main/java/eu/tsalliance/streamer/console/CommandHandler.java
package eu.tsalliance.streamer.console;

import lombok.Getter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import org.reflections.Reflections;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Set;

public class CommandHandler {
    private static final Logger logger = LoggerFactory.getLogger(CommandHandler.class);

    @Getter private static final HashMap<String, Command> commands = new HashMap<>();
    
    public static void registerCommands() {
        try {
            commands.clear();
            Reflections reflections = new Reflections("eu.tsalliance.streamer.commands");
            Set<Class<? extends Command>> classSet = reflections.getSubTypesOf(Command.class);

            classSet.forEach(aClass -> {
                try {
                    Command command = aClass.getConstructor().newInstance();

                    for(String alias : command.getAliases()) {
                        commands.put(alias.toLowerCase(), command);
                    }

                    commands.put(command.getName().toLowerCase(), command);
                } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
                    e.printStackTrace();
                }
            });
        } catch (Exception ignored) {
            logger.error("registerCommands(): Failed registering commands");
        }
    }

    public static void handleCommand(String input) {
        new Thread(() -> {
            ArrayList<String> query = new ArrayList<>(Arrays.asList(input.split(" ")));
            String commandName = query.remove(0).toLowerCase();

            try {
                Command command = getCommands().get(commandName.toLowerCase());

                if(command == null) {
                    throw new NullPointerException();
                }

                command.execute(commandName, query);
            } catch (NullPointerException ignored) {
                logger.error("Command '" + input + "' not found.");
            }
        }).start();
    }
}

File: discordbot/src/main/kotlin/live/tsradio/discordbot/command/handler/CommandHandler.kt
package live.tsradio.discordbot.command.handler

import live.tsradio.discordbot.command.Category
import live.tsradio.discordbot.command.Command
import live.tsradio.discordbot.command.sender.ConsoleSender
import live.tsradio.discordbot.command.sender.DiscordSender
import live.tsradio.discordbot.command.sender.Sender
import live.tsradio.discordbot.config.MainConfig
import live.tsradio.discordbot.language.Lang
import discord4j.core.`object`.entity.Guild
import discord4j.core.`object`.entity.Message
import org.reflections.Reflections
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import kotlin.collections.ArrayList
import kotlin.collections.HashMap

object CommandHandler {
    private val logger: Logger = LoggerFactory.getLogger(CommandHandler::class.java)

    val commands: HashMap<String, Command> = HashMap()
    var activeThreads = 0

    fun registerCommands(){
        try {
            val reflections = Reflections("live.tsradio.discordbot.command.commands")
            val classSet = reflections.getSubTypesOf(Command::class.java)

            classSet.forEach { cmdClass ->
                run {
                    val command: Command = cmdClass.getConstructor().newInstance()
                    commands[command.name] = command
                    logger.info("registerCommands(): Registering '${cmdClass.simpleName}'")
                }
            }

            logger.info("registerCommands(): Registered ${commands.size} " + when(commands.size) {
                1 -> "command successfully."
                else -> "commands successfully."
            })

        } catch (ignored: Exception){
            logger.error("registerCommands(): Failed registering existing commands. The bot may not listen to commands.")
            ignored.printStackTrace()
        }
    }

    fun handleCommand(guild: Guild?, eventMessage: Message?, message: String, sender: Sender) {
        val thread = Thread {
            val content: String = message.removePrefix(MainConfig.getString("general/prefix"))
            val query = ArrayList<String>(content.split(" "))
            val cmdName: String = query.removeAt(0)

            try {
                val command: Command? = commands[cmdName]
                if (command == null || (command.category == Category.HIDDEN && sender is DiscordSender)) {
                    sender.sendError(Lang.getString("error_cmd_not_found").replace("%name%", cmdName)).subscribe()
                } else {
                    if((sender is ConsoleSender) && !command.consoleOptimised) {
                        sender.sendError("This command is not available for console input.")
                    } else {
                        command.execute(sender, eventMessage, guild, query)
                    }
                }
            } catch (ignored: KotlinNullPointerException){ }
            catch (ex: Exception){
                ex.printStackTrace()
            }

            --activeThreads
        }
        ++activeThreads
        thread.name = "bot-commandhandler-$activeThreads"
        thread.start()
    }

}